core/templates/components/question-directives/question-misconception-editor/question-misconception-editor.component.ts

```diff
@@ -16,123 +16,142 @@
  * @fileoverview Component for the question misconception editor.
  */
 
-require('directives/angular-html-bind.directive.ts');
+import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
+import { downgradeComponent } from '@angular/upgrade/static';
+import { NgbModal, NgbModalRef } from '@ng-bootstrap/ng-bootstrap';
+import cloneDeep from 'lodash/cloneDeep';
+import { StateEditorService } from 'components/state-editor/state-editor-properties-services/state-editor.service';
+import { Misconception, MisconceptionSkillMap } from 'domain/skill/MisconceptionObjectFactory';
+import { ExternalSaveService } from 'services/external-save.service';
+import { TagMisconceptionModalComponent } from './tag-misconception-modal-component';
+import { SubtitledHtmlBackendDict } from 'domain/exploration/subtitled-html.model';
 
-require('domain/utilities/url-interpolation.service.ts');
-require(
-  'components/state-editor/state-editor-properties-services/' +
-  'state-editor.service.ts');
-require(
-  'components/question-directives/question-misconception-selector/' +
-  'question-misconception-selector.component.ts');
-require(
-  'components/question-directives/question-misconception-editor/' +
-  'tag-misconception-modal.controller.ts');
-require('services/external-save.service.ts');
+export interface MisconceptionUpdatedValues {
+  misconception: Misconception;
+  skillId: string;
+  feedbackIsUsed: boolean;
+}
 
-angular.module('oppia').component('questionMisconceptionEditor', {
-  bindings: {
-    getOnSaveAnswerGroupFeedbackFn: '&onSaveAnswerGroupFeedback',
-    getOnSaveTaggedMisconception: '&onSaveTaggedMisconception',
-    getTaggedSkillMisconceptionId: '&taggedSkillMisconceptionId',
-    isEditable: '=',
-    outcome: '=',
-    rules: '=',
-  },
-  template: require('./question-misconception-editor.component.html'),
-  controller: [
-    '$uibModal', 'ExternalSaveService', 'StateEditorService',
-    function(
-        $uibModal, ExternalSaveService, StateEditorService) {
-      var ctrl = this;
+interface Outcome {
+  feedback: SubtitledHtmlBackendDict;
+}
 
-      ctrl.containsMisconceptions = function() {
-        var containsMisconceptions = false;
-        Object.keys(ctrl.misconceptionsBySkill).forEach(function(skillId) {
-          if (ctrl.misconceptionsBySkill[skillId].length > 0) {
-            containsMisconceptions = true;
-          }
-        });
-        return containsMisconceptions;
-      };
+@Component({
+  selector: 'oppia-question-misconception-editor',
+  templateUrl: './question-misconception-editor.component.html'
+})
+export class QuestionMisconceptionEditorComponent implements OnInit {
+  @Output() onSaveAnswerGroupFeedback:
+    EventEmitter<object> = (new EventEmitter());
+  @Output() onSaveTaggedMisconception:
+    EventEmitter<object> = (new EventEmitter());
```

Comment:
Drop the `on` prefix. The object is too generic and does not provide information about its properties. Use a interface instead.

Reply:
Done.